accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -1236,18 +1248,92 @@ the `LEGACY_ERASURE` of the types.
+follows. Let `S1` and `S2` be the types to which `T1` and `T2` correspond.  Note
+that by assumption, **NORM**(`S1`) and **NORM**(`S2`) are syntactically
+identical up to the identity of bound variables and the placement of legacy type
+modifiers.  We choose `R` arbitrarily as the type literal corresponding to one
+of **NORM**(`S1`) or **NORM**(`S2`) with all legacy type modifiers erased.
+
+The effect of this definition is to ensure that type literals which look
+identical in the source syntax but which may differ by the presence of legacy
+type modifiers are canonicalized consistently in the sense that any two type
+literals which would compare equal via the definition of runtime type equality
+given above will be canonicalized to the same type object.  The choice of what
+type object to canonicalize to is arbitrary in that placement of legacy
+modifiers in type literals is not otherwise observable in the language.  Any
+consistent choice of placement of legacy modifiers is hence an equally valid
+implementation.
+
+Note that the choice of canonicalization for type literals does not depend
+directly on whether sound or unsound null safety semantics are in use.  However,
+soundly null safe programs will never contain legacy types, and hence the
+erasure of legacy type modifiers is never necessary when running in sound mode.
+
+#### Constant instances
+
+In both sound and unsound null checking, canonicalization of generic object
+instances is performed relative to the normal forms of their generic type
+arguments.  That is, an instance of `C<T0>` canonicalizes to the same object as
+`C<T1>` if `T0` and `T1` have the same normal form (up to the identity of bound
+variables), and the objects are otherwise identical.
 
 With sound null checking, all generic const constructors and generic const
-literals are evaluated using the actual type arguments provided, whether legacy
-or non-legacy.  This ensures that with sound null checking, the final
-consistent semantics are obeyed.
+literals are evaluated using the actual type arguments provided.  This ensures
+that with sound null checking, the final consistent semantics are obeyed.
+
+With unsound null checking, all generic const constructors and generic const
+literals are additionally treated as if all type arguments passed to them were
+legacy types regardless of whether the constructed class is defined in a legacy
+library or not, and regardless of whether the constructor invocation or literal
+occurs in a legacy library or not.  That is, a constructor invocation or object
+literal with generic type parameters `Ti` is treated as if the parameters were
+**LEGACY_TYPE**('Ti'). Implementations may choose to eagerly normalize the type
+arguments before applying the legacy rewrite, as desired.
+
+This ensures that const objects which appear identical in the syntax continue to
+canonicalize consistently across legacy and opted-in libraries.  Unlike the case
```

Comment:
It seems likely that we wish to ensure that a library _L1_ can be opted in, and a different library _L2_ can still be opted out, and _L1_ and _L2_ can contain constant expressions that have the same spelling and are canonicalized to the same object. In other words, they keep working if they are at first opted out, and then _L1_ is switched over to enable null safety.

Why, then, is it not important to avoid the breaking change (in _L1_ as well as _L2_) introduced by normalization in mixed version programs?

If this is because normalization is almost never used anyway then perhaps it is also not important to introduce it as part of canonicalization?

Reply:
(Resolving: Normalization is not a new behavior, we've had it since 2.8.0-dev.17.0, hence there is no breakage).